### PR TITLE
Python 3 and PyPy

### PR DESCRIPTION
**Pull request by jamadden**
This is not yet ready for detailed review or commit, just opening to make sure it builds here and to get the full coverage numbers.

We need zope.app.component to be released, and there's a blocking bug in https://github.com/zopefoundation/zope.pluggableauth/issues/7 (that we have a temporary workaround for).

Also still needs cleanups.

### CODE REVIEW

**Comment by jamadden** [comment]
I think the code here is now in good shape and ready for review. We need zopefoundation/zope.app.component#1 to be released and zopefoundation/zope.pluggableauth#8 to be released to move forward with this.

File: .travis.yml [hunk 1]
```diff
@@ -1,10 +1,28 @@
...
+  - coveralls
 notifications:
-    email: false
+  email: false
+
+install:
+  - pip install -U pip setuptools
+  - pip install -U coveralls coverage
+  # XXX Temporary
+  - pip install git+https://github.com/zopefoundation/zope.app.component@python3#egg=zope.app.component
```
Comment (mgedmin) [inline .travis.yml]: Reminder to remove this when zope.app.component is fixed.

File: setup.py [hunk 2]
```diff
@@ -87,14 +109,14 @@ def read(*rnames):
...
-          'zope.pluggableauth >= 1.1',
+          'zope.pluggableauth >= 2.1',
           'zope.schema',
           'zope.security',
           'zope.traversing',
           # Needed for browser code.
-          'zope.app.container',
-          'zope.app.component',
+          'zope.app.container >= 4.0.0',
+          'zope.app.component >= 4.0.0.dev0',
```
Comment (mgedmin) [inline setup.py]: Reminder to change this to `4.0.0` without `.dev0` when that version is out.

File: src/zope/app/authentication/browser/loginform.py [hunk 3]
```diff
@@ -13,30 +13,35 @@
...
         unauthenticated = IUnauthenticatedPrincipal.providedBy(principal)
         self.unauthenticated = unauthenticated
-        
+
         camefrom = request.get('camefrom')
-        if isinstance(camefrom, list):
+        if isinstance(camefrom, list): # pragma: no cover
             # this can happen on python2.6, as it changed the
             # behaviour of cgi.FieldStorage a bit.
+            # XXX: Just Python 2.6 or later too? Tests don't produce this.
```
Comment (mgedmin) [inline src/zope/app/authentication/browser/loginform.py]: I haven't the energy to verify, but presumably this happens when the HTTP request contains multiple query parameters with the same name -- you could try `GET /@@loginForm.html/?camefrom=foo&camefrom=bar` to test this corner case.

And I believe the behavior is the same for all Python versions >= 2.6.  (TBH I don't remember Python 2.5 behaving differently).

File: src/zope/app/authentication/browser/tests/test_doctests.py [hunk 4]
```diff
@@ -154,5 +176,10 @@ def test_suite():
...
         unittest.makeSuite(FunkTest),
         issue663,
-        doctest.DocFileSuite('../schemasearch.txt'),
+        doctest.DocFileSuite('../schemasearch.rst'),
         ))
+
+# XXX: Temporary workaround for zope.pluggableauth#7
+if str is not bytes:
+    import zope.pluggableauth.plugins.idpicker
+    zope.pluggableauth.plugins.idpicker.unicode = str
```
Comment (mgedmin) [inline src/zope/app/authentication/browser/tests/test_doctests.py]: Reminder to remove this when zope.pluggableauth is fixed.

File: tox.ini [hunk 5]
```diff
@@ -0,0 +1,10 @@
+[tox]
+envlist =
+    py27, pypy, py34, py35, py36
+
+[testenv]
+commands =
+    zope-testrunner --test-path=src []
+deps =
+   git+https://github.com/zopefoundation/zope.app.component@python3#egg=zope.app.component
```
Comment (mgedmin) [inline tox.ini]: Reminder to remove this once z.a.component is fixed.

**Review by mgedmin — APPROVED** [review]
~~tl;dr~~ lgtm

(You're doing enormous amounts of work!  Thank you!)

**Comment by jamadden** [comment]
zope.app.component 4.0 and zope.pluggableauth 2.2 have been released so all deps necessary for this are ready.

As an aside, you'll notice that the principalfolder.rst and special-groups.rst tests got rewritten from sending raw HTTP form POSTs to being Browser based. I did this during the porting process as I found this much easier to debug (because I could look at the expected input HTML step by step). By the time I had those two tests working, the other tests were working too so I left them using raw HTTP.

**Comment by mgedmin** [comment]
FWIW you now have PyPI rights to zope.app.authentication.

File: src/zope/app/authentication/browser/tests/test_doctests.py (hunk 4 above)
Comment (jamadden) [reply src/zope/app/authentication/browser/tests/test_doctests.py]: Thank you, I almost forgot!

File: src/zope/app/authentication/browser/loginform.py (hunk 3 above)
Comment (jamadden) [reply src/zope/app/authentication/browser/loginform.py]: That did it! Excellent, thanks.

**Comment by jamadden** [comment]
Thank you!
